src/wbase/Task.cc

```diff
@@ -91,21 +77,15 @@ bool Task::ChunkEqual::operator()(Task::Ptr const& x, Task::Ptr const& y) {
     if (!x || !y) {
         return false;
     }
-    if ((!x->msg) || (!y->msg)) {
-        return false;
-    }
-    return x->msg->has_chunkid() && y->msg->has_chunkid() && x->msg->chunkid() == y->msg->chunkid();
+    return x->_chunkId == y->_chunkId;
 }
 
 // Task::PtrChunkIdGreater functor
 bool Task::ChunkIdGreater::operator()(Task::Ptr const& x, Task::Ptr const& y) {
     if (!x || !y) {
         return false;
     }
-    if ((!x->msg) || (!y->msg)) {
-        return false;
-    }
-    return x->msg->chunkid() > y->msg->chunkid();
+    return x->_chunkId > y->_chunkId;
```

Comment:
Same comment as for `ChunkEqual` - the comparison should also involve the new attribute `_hasChunkId` which is relevant in this context. Something like this:
```
    if (!x->_hasChunkId || !y->_hasChunkId) {
        throw std::runtime_error("Task::ChunkIdGreater::" + std::string(__func__) + ": incompatible states to allow the comparison");
    }
    return x->_chunkId > y->_chunkId;
```

Reply:
If you don't want to introduce the new attribute `_hasChunkId` and still want to use `-1` then you still need to throw an exception if the states are not comparable. The latter means if either or both objects store `-1` as the chunk identifier. This reinforcement (attribute validation) would lead to a safer code which would allow for detecting issues with chunks at an earlier stage.

Reply:
This is a comparison that above all else, needs to be consistent so items can be sorted. Having it throw an exception would be disaterous to any form of sorting.

What these comparison functions do is sort the items of the lists so they can be compared consistently. 
As long as all tasks without a valid chunk id have a value of -1, the sorting will always be consistent and useful, even though it probably never happens.